zowe/zlux-server-framework
# PoC for Leader Election

## PoC for Leader Election

### New environment variables shared across zlux instances:  
* ZLUX_RAFT_CLUSTER_ENABLED, set to TRUE or YES
* ZOWE_INSTANCE, used in session cookie name
* ZOWE_SESSION_SECRET - temporary solution
* ZLUX_RAFT_CLUSTER_SIZE - number of peers in zlux app server cluster, default - 3
* ZLUX_RAFT_MAX_LOG_SIZE - max Raft log size, default is 100 items
* ZLUX_RAFT_MIN_ELECTION_TIMEOUT - min election timeout in milliseconds, default 1000 ms
* ZLUX_RAFT_MAX_ELECTION_TIMEOUT - max election timeout in milliseconds, default 2000 ms
* ZLUX_RAFT_HEARTBEAT_INTERVAL - heartbeat interval in milliseconds, default 200 ms


### How to test a failover
0. Set global environment variables
```
ZLUX_RAFT_CLUSTER_ENABLED=TRUE
ZOWE_INSTANCE="my_zowe"
ZOWE_SESSION_SECRET="my-secret"
ZLUX_RAFT_CLUSTER_SIZE=3
```
0.  Compile Sample Angular App using branch https://github.com/lchudinov/zowe-sample-angular-app/tree/feature/add-storage-test
1. Setup three zlux instances with Mediation layer enabled
2. Run API Mediation Layer
3. Run all three instances
4. Open Eureka UI at https://localhost:10011 (user/user) and wait until the instances register in Eureka, find each other, and perform Leader Election. 
5. When Leader Election is done you will see one ZLUX instance with status `UP` and the other two instances with status `OUT_OF_SERVICE`, e.g. 
![image](https://user-images.githubusercontent.com/34669449/81289436-dced3180-907f-11ea-9d64-0df9bf16ff1d.png)
6. Open Zlux UI at https://localhost:10010/ui/v1/zlux/ZLUX/plugins/org.zowe.zlux.bootstrap/web/, run Sample Angular App.
7. Add a new key-value pair into a cluster storage using Sample Angular App UI: 
![storage-test](https://user-images.githubusercontent.com/34669449/81290773-145cdd80-9082-11ea-8889-9921f1a20277.gif)
8.  Stop Leader instance
9. Open Eureka UI at https://localhost:10011 and check that a new Leader has been elected: 
![image](https://user-images.githubusercontent.com/34669449/81290972-643ba480-9082-11ea-95f0-4766033758b4.png) Note that old leader might appear in status `OUT_OF_SERVICE` for some time.
7. Ensure that Zlux UI still works. Press `F5`. UI must not show a login dialog because your session has been replicated on all servers.
8. Open Sample Angular App and check that cluster storage has been replicated:
![storage-test-2](https://user-images.githubusercontent.com/34669449/81291438-2e4af000-9083-11ea-990a-a1d0d02386db.gif)
